rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,323 @@
+
+##### Pros:
+
+*   No need for RFC or new public API
+
+##### Cons:
+
+*   Downstream tools are stuck between (a) trying to implement ops, even if the
+    efficiency is not yet a concern. In which case this just redos
+    implementation in the downstream tooling using its own language. Or (b)
+    trying to extract certain graph patterns that can be optimizing and
+    following the moving target. Once such pattern is extracted there is a
+    pressure on op-maintainers to freeze his implementation to avoid breaking
+    the patterns used by the downstream tools. Or (c) Tring to roll out their
+    own extension such as `OpHint` for tflite.
+
+*   New ops that could push the industry forward however are stymied due to lack
+    efficient implementation, software vendors are not interested in
+    implementation until the ops become popular enough, creating a vicious
+    cycle.
+
+#### Alternative 1: Recommend each new op to be implemented as a new atomic node in graphdef.
+
+#### Pros:
+
+*   Simplified graph processing since the operation is just a single Node in the
+    graph.
+*   Clean graph def.
+*   Tooling (once supports new ops) stays forward compatible. The underlying
+    implementation is invisible.
+
+#### Cons:
+
+Introduces backward incompatible changes that could be avoided. Every time an
+operation is added, the versions of Tensorflow that would have been capable of
+processing the original graph (with core ops), will no longer be able to read
+the graph def that contains the new to process the new version, even though the
+only thing that *has* changed is the tensorflow added a faster implementation of
+that Op.
+
+For example: consider matrix exponentiation. It is implemented as a fairly
+complex tensorflow graph that uses just regular matrix multiplication and other
+standard ops. However, one can easily imagine that this implementation could be
+highly optimized if done as a single expm node, however if we replace that, it
+will break old tools, which would either need to effectively re-implement
+tensorflow original implementation.
+
+Requires custom version of tensorflow to add new ops outside of tensorflow which
+makes it out of reach for most users and introduces incompatibilities within the
+ecosystem, effectively forcing large users to be stuck at old versions of tensorflow.
+
+#### Alternative 2: Use name scopes for delineation of the ops
+
+Pros:
+
+*   Simple and intuitive graph-def
+*   Backward compatible - no new operations are added.
+
+Cons:
+
+*   very brittle for the tools to optimize such ops. If they depend on scope
+    names it can easily cause conflicts with models that are doing something
+    else or accidental renames can cause scopes to become invisible for the
+    graph processing.
+*   If tools depend on graph pattern matching, this makes it hard to change
+    implementations later on.
+
+*   Tooling is not forward compatible.
+
+## Questions and Discussion Topics
```

Comment:
Interesting question. My intent was to leave it  up to tooling to decide what they want to do. In this sense the field is leveled for both core `tensorflow` and downstream tools (such as tf-lite).  if `tensorflow` has an efficient implementation it can simply use it, or it can rely on function implementation. Or it can try both and decide what is better at runtime. Basically it brings what tensorflow already does with conv2d to the api level. Or perhaps i misread this comment.  Do you have some specific suggestion on specialization in mind?

Reply:
Sorry if my initial comment has been unclear. Specialization itself will have to be implemented in downstream tools. I was actually thinking about implementation selector in Grappler while writing that. However, to allow that, an op should somehow indicate that it's targeting a specific device, or is device agnostic.

Reply:
I see, if i understand it right: basically the idea is that different block implementations  could  be more efficient on different target devices and at  grappler (or downstream tooling for that matter) should be able to put the correct implementation in based on the device it is targetting.  I certainly can see a utility for doing that though i think will need to think how to avoid it being very messy.  Will add this to discussion list.

Reply:
@qlzh727 I think your specialization work should commute with this, can you comment?